Use replace_resource only if resource already exists

knatsu-summer opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/IBM/oper8/blob/main/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

## Related Issue
Supports #ISSUE_NUMBER

## Related PRs
This PR is not dependent on any other PR

## What this PR does / why we need it
Encountering a 404 error when user uses `_replace_resource` option to deploy a resource that is not currently in the environment.
I would like to add logic to use `_apply_resource` instead of `_replace_resource` when first deploying.

## Special notes for your reviewer

## If applicable**
- [ ] this PR contains documentation
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

## What gif most accurately describes how I feel towards this PR?
![Example of a gif](https://media.giphy.com/media/snwvCcEKk33Hy/giphy.gif)

HonakerM reviewed (approved):
LGTM!

gabe-l-hart reviewed (changes requested):
Thanks for the fix! It looks like this PR needs a couple of things before it's merge ready:

1. Rebase with signoff to get DCO passing (`git fetch upstream && git rebase upstream/main --signoff && git push -f`)
2. Get the unit tests passing

knatsu-summer commented:
@gabe-l-hart @HonakerM 
The unit tests are still failing. 
The failure comes from [this line](https://github.com/knatsu-summer/oper8/blob/f7ceb8b99975fa0f5ed5d6af015718e79ef81415/tests/deploy_manager/test_openshift_deploy_manager.py#L244)
```
>       assert changed
E       assert False
```
I tested it locally and added some log prints. It seems that the resource is not being updated after the replacement. The data and path used for the replacement (PUT method) are:
I tested it locally and added some log prints., it seems that the [resource](https://github.com/knatsu-summer/oper8/blob/f7ceb8b99975fa0f5ed5d6af015718e79ef81415/oper8/deploy_manager/openshift_deploy_manager.py#L948) is not being updated after `replace` function.
The data and path used for `replace` (PUT method):

- **request body:**

```
{
    "kind": "Foo",
    "apiVersion": "foo.bar.com/v1",
    "metadata": {
        "name": "bar",
        "namespace": "test",
        "managedFields": null
    },
    "spec": {
        "some": "key_1"
    }
}
```

- **request path:** 

`/apis/foo.bar.com/v1/namespaces/test/foos/bar`

The values of [`current` and `apply_res`](https://github.com/knatsu-summer/oper8/blob/f7ceb8b99975fa0f5ed5d6af015718e79ef81415/oper8/deploy_manager/openshift_deploy_manager.py#L948) are the same as follows::
```
current: {'apiVersion': 'foo.bar.com/v1', 'kind': 'Foo', 'metadata': {'name': 'bar', 'namespace': 'test', 'managedFields': None}}, 
apply_res:{'apiVersion': 'foo.bar.com/v1', 'kind': 'Foo', 'metadata': {'name': 'bar', 'namespace': 'test', 'managedFields': None}}
```
I tried `server_side_apply` and it works as expected. 
Do you have any thoughts?

gabe-l-hart commented:
@knatsu-summer this found a pretty gnarly bug in the `KubMock` implementation of `PUT`. I've pushed a commit that should get the tests in line.

gabe-l-hart reviewed (approved):
With the fix in `kub_mock`, this should be good to go!
